### ISSUE TITLE
FastPitch trained from scratch nemo 2.5.3: audio quality degrades after ~10 seconds

### ISSUE THREAD

**Issue by stefanosalvucci** [NVIDIA-NeMo/NeMo#15140]
Hello, I trained FastPitch from scratch with nemo docker image 2.5.3 on an Italian dataset ([z-uo/male-LJSpeech-italian](https://huggingface.co/datasets/z-uo/male-LJSpeech-italian)) using the attached config, then trained HiFi-GAN as vocoder.

Problem: Generated audio quality is good for the first ~10 seconds, then noticeably degrades. For example, an 13-second generation has clean audio for the first 10 seconds, but the remaining 3 seconds sound distorted/garbled.

Test: attached a .zip with an ok audio and an audio with last 5 seconds degraded

i run the train with
python3 fastpitch.py
–config-name=fastpitch_align_v1.05.yaml
model.train_ds.dataloader_params.batch_size=32
model.validation_ds.dataloader_params.batch_size=32
train_dataset=./italian_tts_data/manifest_train.json
validation_datasets=./italian_tts_data/manifest_val.json
sup_data_path=./italian_tts_data/sup_data
exp_manager.exp_dir=./results

Has anyone experienced the same issue? Any insights on what might be causing this?

[test-audio.zip](https://github.com/user-attachments/files/23903504/test-audio.zip)

[config_fastpitch.txt](https://github.com/user-attachments/files/23903523/config_fastpitch.txt)

Thanks a lot

**Comment by stefanosalvucci** [comment]
<p style="margin-top: 0px; color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong style="font-weight: bolder;">Update — additional tests on RTX 4060 and RTX 4090</strong></p><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">I ran two additional FastPitch-from-scratch trainings on the same Italian dataset, using identical training commands and almost identical configs.</p><h3 style="font-variation-settings: normal; font-feature-settings: normal; font-family: Arial, sans-serif; margin: 2rem 0px 0.67rem; line-height: 1.2; font-size: 17.2305px; color: rgb(51, 51, 51); font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a name="p-1726175-hardware-environment-differences-1" class="anchor" href="https://forums.developer.nvidia.com/t/fastpitch-trained-from-scratch-audio-quality-degrades-after-10-seconds/353329#p-1726175-hardware-environment-differences-1" style="background-color: rgba(0, 0, 0, 0); color: rgb(118, 185, 0); text-decoration: none; cursor: pointer; overflow-wrap: break-word; opacity: 0; transition: opacity 0.25s;"></a><strong style="font-weight: bolder;">Hardware / environment differences</strong></h3><div class="md-table fullscreen-table-wrapper --editable" data-table-index="0" style="overflow-y: auto; margin: 1em 0px; display: block; position: relative; transition: box-shadow 0.6s cubic-bezier(0.165, 0.84, 0.44, 1); color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="fullscreen-table-wrapper__buttons" style="position: absolute; top: 0.5rem; right: 0.5rem;"><button class="open-popup-link btn-default btn btn-icon no-text btn-edit-table" title="Edit Table" style="font-family: inherit; font-size: 1em; line-height: normal; margin: 0px; overflow: visible; text-transform: none; appearance: auto; color: rgb(51, 51, 51); cursor: pointer; font-variation-settings: inherit; font-feature-settings: inherit; font-weight: bold; background: rgba(0, 0, 0, 0); transition: opacity 0.25s ease-in-out; border-color: rgb(153, 153, 153); border-style: solid; border-width: 1px; border-image: none 100% / 1 / 0 stretch; display: inline-block; align-items: center; justify-content: center; box-sizing: border-box; padding: 11px 12px; border-radius: 4px; margin-inline: 0.25em; position: sticky; left: 1rem; opacity: 0; white-space: nowrap; user-select: none;"><svg class="fa d-icon d-icon-pencil svg-icon edit-table-icon svg-node" aria-hidden="true"></svg></button></div>


<img width="597" height="117" alt="Image" src="https://github.com/user-attachments/assets/389c03ea-091c-41ac-9c07-fac3cba72728" />

</div><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">All other configuration parameters (including pitch_mean, pitch_std, fmin, fmax) and the training command were the same.</p><h3 style="font-variation-settings: normal; font-feature-settings: normal; font-family: Arial, sans-serif; margin: 2rem 0px 0.67rem; line-height: 1.2; font-size: 17.2305px; color: rgb(51, 51, 51); font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a name="p-1726175-observed-audio-differences-2" class="anchor" href="https://forums.developer.nvidia.com/t/fastpitch-trained-from-scratch-audio-quality-degrades-after-10-seconds/353329#p-1726175-observed-audio-differences-2" style="background-color: rgba(0, 0, 0, 0); color: rgb(118, 185, 0); text-decoration: none; cursor: pointer; overflow-wrap: break-word; opacity: 0; transition: opacity 0.25s;"></a><strong style="font-weight: bolder;">Observed audio differences</strong></h3><ul style="margin: 1em 0px 1em 1.25em; padding: 0px; clear: both; padding-inline-start: 1.25em; color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li><p>The<span> </span><strong style="font-weight: bolder;">4060ti (NeMo 2.2.0)</strong><span> </span>model produces<span> </span><strong style="font-weight: bolder;">more stable audio</strong>:<br>degradation still happens after ~20 seconds (on a 28s audio), but it is noticeably milder.</p></li><li><p>The<span> </span><strong style="font-weight: bolder;">4090 (NeMo 2.2.1)</strong><span> </span>model produces<span> </span><strong style="font-weight: bolder;">much worse long-form degradation</strong>:<br>the first ~15 seconds sound correct, but the last part of the generation becomes heavily distorted/garbled, audio is sensibly slower than the one in 4060</p></li><li><p>same text input, on 4060 generates 28s audio, on 4090 produces 34s audio</p></li><li><p>Both models were synthesized using the same HiFi-GAN (<code style="font-family: &quot;JetBrains Mono&quot;, Consolas, Monaco, monospace; font-size: 14px; color: rgb(51, 51, 51); background: none 0% 0% / auto repeat scroll padding-box border-box rgb(243, 243, 243); border-radius: 4px; line-height: 1.30769; padding: 2px 4px; white-space: pre-wrap; max-width: 1100px;">tts_en_lj_hifigan_ft_mixerttsx</code>).</p></li></ul><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Even more:<br><strong style="font-weight: bolder;">we trained also fastpitch with english ljspeech dataset (the same EN dataset used for the EN pretrained model) and the model shows a similar small quality drop at the end of long-form generations</strong>, which suggests this might not be dataset-specific.</p><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">I’ve attached two new audio samples comparing the 4060ti and 4090 outputs on the same input text.</p><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">If anyone has seen similar behavior or has suggestions on what to check (dataset pitch stats, config simplification, NeMo version pinning, long-form TTS stability issues, etc.), any insight would be super helpful.</p><p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

[tests-4060ti-4090.zip](https://github.com/user-attachments/files/23928453/tests-4060ti-4090.zip)

<p style="color: rgb(51, 51, 51); font-family: &quot;NVIDIA Sans&quot;, Helvetica, Arial, sans-serif; font-size: 15px; font-style: normal; font-variant-ligatures: none; font-variant-caps: normal; font-weight: 400; letter-spacing: 0.15px; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Thanks again!</p>

**Comment by MuyangDu** [comment]
Hi Stefano, FastPitch uses absolute positional embeddings in both its encoder and decoder, which limits its ability to generalize beyond the maximum audio length seen during training (though it may handle slightly longer sequences). A simple workaround is to split the input text into shorter sentences, synthesize them separately, and then concatenate the resulting audio segments. If you need the model to directly generate very long speech, you will need to train it on data with longer durations.

**Comment by pzelasko** [comment]
CC @blisc @rlangman
